scripts/zpool

```diff
@@ -47,28 +47,40 @@ This is a resource agent that manages ZFS zpools.
 EOF
 }
 
+zpool_status_log() {
+    # Log zpool status to splunk for post-mortem.  Run in it the background
+    # in case zpool status hangs.  Wait two seconds for it to log.
+    nohup sh -c 'zpool status | logger -t resource_agent_fail -p local7.debug' &
+    sleep 2
+}
+
 zpool_monitor() {
     local rc
+    rc=$OCF_NOT_RUNNING
 
     # exit immediately if configuration is not valid
     zpool_validate_all || exit $?
 
-    ocf_run -q -info zpool status $OCF_RESKEY_pool
+    if [ -e /proc/spl/kstat/zfs/$OCF_RESKEY_pool/state ] ; then
+        # Best case, we can read the pool health locklessly from the proc file.
+        state=$(cat /proc/spl/kstat/zfs/$OCF_RESKEY_pool/state)
 
-    # This example assumes the following exit code convention
-    # for zpool status:
-    # 0: imported
-    # any other: not imported
-    case "$?" in
-        0)
+        if [ "$state" == "SUSPENDED" ] || [ "$state" == "FAULTED" ] || [ "$state" == "UNAVAIL" ] ; then
+            ocf_log debug "ZFS pool $OCF_RESKEY_pool is $state"
+            zpool_status_log
+
+            # There's no good error code that fits here.  OCF_ERR_ARGS is
+            # the least bad choice: "The resource's configuration is not valid
+            # on this machine".
+            rc=$OCF_ERR_ARGS
```

Comment:
OCF_ERR_GENERIC is a soft error, and we want to return a hard error (https://clusterlabs.org/pacemaker/doc/en-US/Pacemaker/1.1/html/Pacemaker_Explained/_how_are_ocf_return_codes_interpreted.html).

I'll update it with this comment:
```            
# There's no good error code that fits here.  We need a "hard"
# error to move the resource and prevent it from being retried on
# the current node.  OCF_ERR_ARGS is the least bad hard error:
# "The resource's configuration is not valid on this machine".
```

Reply:
I understand how you are interpreting that.  The documentation is certainly not the most clear.  I still think that OCF_ERR_GENERIC is the correct return code.  I believe that when they say "prevent it from being retried on the current node" they _really mean_ prevent it from ever being used on this node again.  OCF_ERR_ARGS really denotes a permanent configuration problem that can never be resolved until the resource agent itself is reconfigured.  But the problem you are dealing with is _temporary_ problem that does _not_ require that the agent be reconfigured.  Someone just needs to debug the underlying issue and then we will want pacemaker to move the resource back to its original home.  It may refuse to do that if we use OCF_ERR_ARGS.

My interpretation seems to be consistent with how the other resource agents in the official clusterlabs resource agents package use the error codes.

Reply:
I'll retest with OCF_ERR_GENERIC.  My concern was that pacemaker would try to restart the pool on the bad node again, since soft errors = "Restart the resource or move it to a new location".  But maybe they mean that the pool will get moved, and I have the option to restart it on the node once the error is resolved.

Reply:
I retested, and saw that `OCF_ERR_GENERIC` just causes pacemaker to restart the resource on the failed node:
```
crmd[156216]:   notice: State transition S_IDLE -> S_POLICY_ENGINE
pengine[156215]:  warning: Processing failed op monitor for jet19_zpool on jet19: unknown error (1)
pengine[156215]:   notice:  * Recover    jet19_zpool        (          jet19 )
pengine[156215]:   notice:  * Restart    lquake-OST0002     (          jet19 )   due to required jet19_zpool start
pengine[156215]:   notice: Calculated transition 274, saving inputs in /var/lib/pacemaker/pengine/pe-input-664.bz2
pengine[156215]:  warning: Processing failed op monitor for jet19_zpool on jet19: unknown error (1)
pengine[156215]:   notice:  * Recover    jet19_zpool        (          jet19 )
pengine[156215]:   notice:  * Restart    lquake-OST0002     (          jet19 )   due to required jet19_zpool start
crmd[156216]:   notice: Initiating stop operation lquake-OST0002_stop_0 locally on jet19
pengine[156215]:   notice: Calculated transition 275, saving inputs in /var/lib/pacemaker/pengine/pe-input-665.bz2
dhcpd[191059]: DHCPDISCOVER from a4:bf:01:54:fc:df via pub: network 134.9.59.0/24: no free leases
dhcpd[191059]: DHCPDISCOVER from a4:bf:01:36:41:4f via pub: network 134.9.59.0/24: no free leases
crmd[156216]:   notice: Result of stop operation for lquake-OST0002 on jet19: 0 (ok)
crmd[156216]:   notice: Initiating stop operation jet19_zpool_stop_0 locally on jet19
crmd[156216]:   notice: Result of stop operation for jet19_zpool on jet19: 0 (ok)
crmd[156216]:   notice: Initiating start operation jet19_zpool_start_0 locally on jet19
```
`OCF_ERR_ARGS` moves the pool to the partner node:
```
crmd[156216]:   notice: State transition S_IDLE -> S_POLICY_ENGINE
pengine[156215]:  warning: Processing failed op monitor for jet19_zpool on jet19: invalid parameter (2)
pengine[156215]:    error: Preventing jet19_zpool from re-starting on jet19: operation monitor failed 'invalid parameter' (2)
pengine[156215]:   notice:  * Recover    jet19_zpool        ( jet19 -> jet20 )
pengine[156215]:   notice:  * Move       lquake-OST0002     ( jet19 -> jet20 )
pengine[156215]:   notice: Calculated transition 276, saving inputs in /var/lib/pacemaker/pengine/pe-input-666.bz2
pengine[156215]:  warning: Processing failed op monitor for jet19_zpool on jet19: invalid parameter (2)
pengine[156215]:    error: Preventing jet19_zpool from re-starting on jet19: operation monitor failed 'invalid parameter' (2)
pengine[156215]:   notice:  * Recover    jet19_zpool        ( jet19 -> jet20 )
pengine[156215]:   notice:  * Move       lquake-OST0002     ( jet19 -> jet20 )
crmd[156216]:   notice: Initiating stop operation lquake-OST0002_stop_0 locally on jet19
```
So I think we want `OCF_ERR_ARGS` or some other hard error here.

Reply:
I don't contest what you saw, but I think you are using the wrong mechanism to achieve the goal.  Decisions about how many times to try before failing over should be handled in pacemaker configuration, not by abusing error codes in the resource agent script.  Here is a better way to adjust pacemaker's behavior if we don't like the defaults:

https://clusterlabs.org/pacemaker/doc/en-US/Pacemaker/1.1/html/Pacemaker_Explained/s-failure-migration.html

Reply:
You don't have to "make it configurable", it is already a configurable in pacemaker.  It is just how pacemaker works.  We can use one meta-attribute to set the migration-threshold for all zpool resource types.

Another pacemaker configuration option to consider is the "on-fail" one.  This option might be a better one in this situation.  on-fail defaults to "restart", but we probably want it to be "fence".  It might even be possible to add 'on-fail="fence"' to the monitor action's defaults in the resource agent script's zpool_meta_data() so that it doesn't have to be set manually.

Reply:
Actually, I take that back about using "fence".  We don't want to fence a node and kill all resources running there just because a single resource failed.

If we want the zpool resource to migrate after only a single failure, then setting migration-threshold=1 for zpool resources seems like the pacemaker way to achieve that goal.

Reply:
Sorry, I wasn't clear - by configurable I mean make it a soft error so we can use `migration-threshold`.  `migration-threshold` isn't applicable to hard or fatal errors.

I still prefer to define this as a hard error, rather than a soft error with `migration-threshold=1`.  They have the same behaviour, but with a hard error, the user can never shoot themselves in the foot by setting `migration-threshold!=1`.   I also would argue that a suspended pool better fits the definition of a hard error than a soft error, since a suspended pool is non-transient (https://clusterlabs.org/pacemaker/doc/en-US/Pacemaker/1.1/html/Pacemaker_Explained/_how_are_ocf_return_codes_interpreted.html).

`on-fail` describes what to do if an operation fails, not what to do when the monitoring function returns an error code.  I tried it by returning a soft error with `on-fail=fence` and it just restarted the resource.  I then tried it with  `on-fail=fence` and `migration-threshold=1` and it just moved the resource.  The node never got fenced.  I believe `on-fail` would kick in if an operation actually failed, like if you couldn't export the pool during a migration.

Reply:
I still think it is an abuse of the error codes to do that, whereas migration-threshold is exactly designed to handle this situation.  Answer me this: With a hard error, how do we do fail back?  With OCF_ERR_ARG, pacemaker thinks that the arguments to the instance are incorrect and can _never_ work on that particular node.  But that isn't really the case.   This is transient problem that will eventually be resolved by a sysadmin, and we will want to put the node back into service without any jumping through hoops.  I think there really is a reason that there is not an OCF_ERR_GENERIC_HARD, and that is because they are not intended to be used this way.

Reply:
> Answer me this: With a hard error, how do we do fail back?

`pcs resource cleanup <resource>`

How about this - I'll update my patch to make it `OCF_ERR_GENERIC` with `migration-threshold=1`, but if pacemaker ever has an `OCF_ERR_MY_POOL_IS_TOTALLY_HOSED`, then we'll consider using that instead :-)   Deal?